### PR TITLE
fix: Live query throws error when constraint `notEqualTo` is set to `null`

### LINKED ISSUE #8834: [Live Query] `matchesKeyConstraints` throwing error when `constraints` is `{$ne : null}`
### New Issue Checklist

- [ x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [ x] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [ x] I have searched through [existing issues](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).
- [ x] I can reproduce the issue with the [latest version of Parse Server](https://github.com/parse-community/parse-server/releases). 

### Issue Description
matchesKeyConstraints function is not handling not equal to null constraint properly. 
Due to which error is thrown `Cannot read properties of null (reading '__type')

image

### Steps to reproduce
Create a subscription on a class with constraint `notEqualTo(key,null)`

### Actual Outcome
Throws error.

### Expected Outcome
Shouldn't throw an error.

### Environment
18.19.0

Server
- Parse Server version: `5.5.5`
- Operating system: `Mac`
- Local or remote host (AWS, Azure, Google Cloud, Heroku, Digital Ocean, etc): `self hosted`

Database
- System (MongoDB or Postgres): `Mongo`
- Local or remote host (MongoDB Atlas, mLab, AWS, Azure, Google Cloud, etc): `Local`

Client
- SDK (iOS, Android, JavaScript, PHP, Unity, etc): `Javascript`
- SDK version: `4.0.1`

### PR DESCRIPTION
**Pull request by RahulLanjewar93**
failing Test case for issue #8834 
adding this so pr is not closed. closes #8834 

image

### CODE REVIEW

**Review by mtrezza — CHANGES_REQUESTED** [review]
Please revert the changes in package-lock.json

**Review by mtrezza — CHANGES_REQUESTED** [review]
This test does not expect anything, please add a check to the test

**Comment by RahulLanjewar93** [comment]
> This test does not expect anything, please add a check to the test

Actually, the server itself crashes since we are not handling the error thrown inside `matchesKeyConstraints`. 
Which is why i have not added any `expect`

File: spec/ParseLiveQuery.spec.js [hunk 1]
```diff
@@ -1269,4 +1269,25 @@ describe('ParseLiveQuery', function () {
...
+
+    const query = new Parse.Query(TestObject);
+    query.notEqualTo('foo', null);
+    await query.subscribe();
+
+    // Create a object and save it, this passes but server crashes afterwards
+    const object1 = new TestObject();
+    object1.set('foo', 'bar');
+    await object1.save();
+    done();
```
Comment (mtrezza) [inline spec/ParseLiveQuery.spec.js]: Remove the `done();` and instead add an expect. The code subscribes to a query, so the expect should probably expect the event to be triggered. Note that we have generally moved away from `done` in favor of `expect`, even tough you may still find it in old tests.

**Review by mtrezza — CHANGES_REQUESTED** [review]

File: spec/ParseLiveQuery.spec.js (hunk 1 above)
Comment (RahulLanjewar93) [reply spec/ParseLiveQuery.spec.js]: have added it. thanks for note.

File: spec/ParseLiveQuery.spec.js [hunk 2]
```diff
@@ -1280,14 +1280,26 @@ describe('ParseLiveQuery', function () {
...
 
+    // Creating a spy so we can register it as event listener.
+    const spy = {
+      create(obj) {
+        expect(obj.attributes.foo).toEqual('bar');
+      },
+    };
+    const createSpy = spyOn(spy, 'create');
+
+    // Subscribe to TestObject class. where foo is not equal to null
```
Comment (mtrezza) [inline spec/ParseLiveQuery.spec.js]:
```suggestion
    const spy = {
      create(obj) {
        expect(obj.attributes.foo).toEqual('bar');
      },
    };
    const createSpy = spyOn(spy, 'create');
    
```

File: spec/ParseLiveQuery.spec.js [hunk 3]
```diff
@@ -1280,14 +1280,26 @@ describe('ParseLiveQuery', function () {
...
     const query = new Parse.Query(TestObject);
     query.notEqualTo('foo', null);
-    await query.subscribe();
+    const subscription = await query.subscribe();
+    subscription.on('create', spy.create);
 
-    // Create a object and save it, this passes but server crashes afterwards
+    // Create a object and save it, the save call passes but server crashes afterwards
```
Comment (mtrezza) [inline spec/ParseLiveQuery.spec.js]:
```suggestion
```

File: spec/ParseLiveQuery.spec.js [hunk 4]
```diff
@@ -1280,14 +1280,26 @@ describe('ParseLiveQuery', function () {
...
     const object1 = new TestObject();
     object1.set('foo', 'bar');
     await object1.save();
-    done();
+
+    // Failing test.Since server crashed, it should not have been called.
```
Comment (mtrezza) [inline spec/ParseLiveQuery.spec.js]:
```suggestion
```

File: spec/ParseLiveQuery.spec.js [hunk 5]
```diff
@@ -1269,4 +1269,32 @@ describe('ParseLiveQuery', function () {
     expect(object2.id).toBeDefined();
     expect(object3.id).toBeDefined();
   });
+
+  it('matchesKeyConstraints fails when subscribing a query with constraint notEqualTo null', async () => {
```
Comment (mtrezza) [inline spec/ParseLiveQuery.spec.js]:
```suggestion
  it('subscribes to query event with constraint not equal to null', async () => {
```

File: spec/ParseLiveQuery.spec.js [hunk 6]
```diff
@@ -1269,4 +1269,32 @@ describe('ParseLiveQuery', function () {
     expect(object2.id).toBeDefined();
     expect(object3.id).toBeDefined();
   });
+
+  it('subscribes to query event with constraint not equal to null', async () => {
```
Comment (mtrezza) [inline spec/ParseLiveQuery.spec.js]:
```suggestion
  it('triggers to query event with constraint not equal to null', async () => {
```

File: spec/ParseLiveQuery.spec.js [hunk 7]
```diff
@@ -1269,4 +1269,32 @@ describe('ParseLiveQuery', function () {
     expect(object2.id).toBeDefined();
     expect(object3.id).toBeDefined();
   });
+
+  it('triggers to query event with constraint not equal to null', async () => {
```
Comment (mtrezza) [inline spec/ParseLiveQuery.spec.js]:
```suggestion
  it('triggers query event with constraint not equal to null', async () => {
```

File: spec/ParseLiveQuery.spec.js [hunk 8]
```diff
@@ -1269,4 +1269,32 @@ describe('ParseLiveQuery', function () {
...
+      silent: true,
+    });
+
+    const spy = {
+      create(obj) {
+        expect(obj.attributes.foo).toEqual('bar');
+      },
+    };
+    const createSpy = spyOn(spy, 'create');
+
```
Comment (mtrezza) [inline spec/ParseLiveQuery.spec.js]:
```suggestion

```

**Comment by mtrezza** [comment]
I believe the test looks good now; do you have any suggestion for a fix?

**Comment by RahulLanjewar93** [comment]
> I believe the test looks good now; do you have any suggestion for a fix?

https://github.com/parse-community/parse-server/blob/2c2563ebd2ac5255b387296a563a34157ec076f5/src/LiveQuery/QueryTools.js#L226

this line should have optional chaining in my opinion

**Review by mtrezza — APPROVED** [review]
Looks good, waiting for CI

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [7.1.0-alpha.10](https://github.com/parse-community/parse-server/releases/tag/7.1.0-alpha.10)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [7.1.0-beta.1](https://github.com/parse-community/parse-server/releases/tag/7.1.0-beta.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [7.1.0](https://github.com/parse-community/parse-server/releases/tag/7.1.0)

**Pull request merged**